### PR TITLE
Use Config class to store credentials

### PR DESCRIPTION
**Pull request by mtczerwinski**
Related to #173.
A user creates a file such as:
```json
{
  "client_id": "xxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.apps.googleusercontent.com",
  "client_secret": "xxxxxxxxxxxxxxxxxxxxxxxx",
  "redirect_uris": [
    "urn:ietf:wg:oauth:2.0:oob"
  ],
  "scope": [
    "https://www.googleapis.com/auth/drive ",
    "https://spreadsheets.google.com/feeds/"
  ],
  "access_token": ""
}
```
which contains everything that should have in order to save a session and allows the user to fetch Google Docs or Sheets.

It requires a little bit more testing (please check it also) and I am waiting for @gimite's opinion and community.

### CODE REVIEW

File: lib/google_drive.rb [hunk 1]
```diff
@@ -94,25 +95,15 @@ def self.saved_session(path = nil, proxy = nil, client_id = nil, client_secret =
...
-        case lines.size
-          when 1
-            token_data = JSON.parse(lines[0].chomp())
-          when 2
-            # Old format.
-            token_data = nil
-          else
-            raise(ArgumentError, "Not a token file: %s" % path)
-        end
+      if config.access_token_exists?
```
Comment (mtczerwinski) [inline lib/google_drive.rb]: or `token_data ||= config.access_token`?

**Comment by gimite** [comment]
Thanks for the pull request. I will take a look.

File: README.rdoc [hunk 2]
```diff
@@ -25,16 +25,15 @@ Example to read/write files in Google Drive:
   require "google_drive"
 
   # The client ID and client secret you obtained in the step above.
-  CLIENT_ID = ...
-  CLIENT_SECRET = ...
+  Keep them in a config file, see an example in the repository.
```
Comment (gimite) [inline README.rdoc]: Where is the example? Maybe you can directly put an example here.

File: README.rdoc [hunk 3]
```diff
@@ -25,16 +25,15 @@ Example to read/write files in Google Drive:
...
 
   # Creates a session. This will prompt the credential via command line for the
-  # first time and save it to ./stored_token.json file for later usages.
+  # first time and save it to ./config.json file for later usages.
   #
   # If you are developing a Web app, and you want to ask the user to log in in
   # the Web app instead of via command line, follow the example code in:
   # http://gimite.net/doc/google-drive-ruby/GoogleDrive.html#method-c-login_with_oauth
-  session = GoogleDrive.saved_session("./stored_token.json", nil, CLIENT_ID, CLIENT_SECRET)
+  session = GoogleDrive.saved_session("./config.json", nil)
```
Comment (gimite) [inline README.rdoc]: Omit the second argument.

File: lib/google_drive.rb (hunk 1 above)
Comment (gimite) [inline lib/google_drive.rb]: Would be fine with:

    token_data = config.access_token

Also, call it refresh_token, neither token_data nor access_token. There is something different called "access token" in OAuth, so calling it access_token is confusing.

File: lib/google_drive.rb [hunk 4]
```diff
@@ -122,34 +113,24 @@ def self.saved_session(path = nil, proxy = nil, client_id = nil, client_secret =
...
-      auth.scope = [
-          "https://www.googleapis.com/auth/drive",
-          "https://spreadsheets.google.com/feeds/"
-      ]
-      auth.redirect_uri = "urn:ietf:wg:oauth:2.0:oob"
-
-      if token_data
+      auth.client_id = config.client_id
+      auth.client_secret = config.client_secret
+      auth.scope = config.scope.join(' ')
```
Comment (gimite) [inline lib/google_drive.rb]: You can specify an array here, so you don't need to join.

Also, please use:

    ["https://www.googleapis.com/auth/drive", "https://spreadsheets.google.com/feeds/"]

as the default when this is omitted in the config file.

File: lib/google_drive.rb [hunk 5]
```diff
@@ -122,34 +113,24 @@ def self.saved_session(path = nil, proxy = nil, client_id = nil, client_secret =
...
+      auth.redirect_uri = config.redirect_uris.first
```
Comment (gimite) [inline lib/google_drive.rb]: Do you know of any use case where you want to specify something other than "urn:ietf:wg:oauth:2.0:oob" here? Otherwise can you just hard-code "urn:ietf:wg:oauth:2.0:oob" as in the original code?

The command line authentication in the else-clause at line 124 only works with "urn:ietf:wg:oauth:2.0:oob". If you directly specify refresh_token in the config, then you don't need to provide redirect_uri here.

File: lib/google_drive.rb [hunk 6]
```diff
@@ -86,7 +86,8 @@ def self.login_with_oauth(client_or_access_token, proxy = nil)
     # +path+ defaults to ENV["HOME"] + "/.ruby_google_drive.token".
     #
     # You can specify your own OAuth +client_id+ and +client_secret+. Otherwise the default one is used.
-    def self.saved_session(path = nil, proxy = nil, client_id = nil, client_secret = nil)
+    def self.saved_session(path = nil, proxy = nil)
```
Comment (gimite) [inline lib/google_drive.rb]: Please keep client_id and client_secret here, and use them instead of those in the config file when given. Otherwise it will break existing code using these parameters.

File: lib/google_drive.rb [hunk 7]
```diff
@@ -122,34 +113,24 @@ def self.saved_session(path = nil, proxy = nil, client_id = nil, client_secret =
...
-      auth.redirect_uri = "urn:ietf:wg:oauth:2.0:oob"
-
-      if token_data
+      auth.client_id = config.client_id
+      auth.client_secret = config.client_secret
+      auth.scope = config.scope.join(' ')
+      auth.redirect_uri = config.redirect_uris.first
 
-        auth.refresh_token = token_data["refresh_token"]
+      unless token_data.nil?
```
Comment (gimite) [inline lib/google_drive.rb]: I prefer "if token_data" here. Double-negation (unless ....nil?) is hard to read.

File: lib/google_drive_v0.rb [hunk 8]
```diff
@@ -6,143 +6,126 @@
 require "json"
 
 require "google_drive_v0/session"
-
+require "shared/config"
```
Comment (gimite) [inline lib/google_drive_v0.rb]: Sorry for taking time to modify this file, but these "v0" version no longer works, so you can just leave it. Can you revert the change for this file? I should delete these files...

File: lib/shared/config.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
+require 'json'
+
+require_relative 'errors/file_not_found'
+
+module GoogleDrive
```
Comment (gimite) [inline lib/shared/config.rb]: Put this file under "lib/google_drive" directory. I try to make file path and class hierarchy (GoogleDrive::Config) consistent.

File: lib/google_drive.rb (hunk 5 above)
Comment (mtczerwinski) [inline lib/google_drive.rb]: As I remember, I generated this file from Google Developer Console and there was a default these values. But as you said I can hard-code it. It is not a problem.

File: lib/shared/config.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  class Config
+    attr_accessor :client_id, :client_secret, :scope, :redirect_uris, :access_token
+
+    def initialize
+      return unless valid?
```
Comment (gimite) [inline lib/shared/config.rb]: I prefer to omit this check, and just let load_config_file throw a standard SystemCallError when the file is not found, rather than defining and throwing our own exception class.

File: lib/shared/config.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+      load_config_file
+    end
+
+    def save
+      ::File.open(file_path,'w') { |file| file.write(prepare_json) }
```
Comment (gimite) [inline lib/shared/config.rb]: Put a whitespace after a comma.

File: lib/shared/config.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
+require 'json'
+
+require_relative 'errors/file_not_found'
+
+module GoogleDrive
+  class Config
```
Comment (gimite) [inline lib/shared/config.rb]: Add

    #:nodoc:

at the end of this line to suppress generating RDoc for this internal class.

File: lib/shared/config.rb [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+      hash.to_json
+    end
+
+    def valid?
+      raise GoogleDrive::FileNotFound.new unless ::File.exist?(file_path)
+      true
+    end
+
+    def file_path
+      ::File.expand_path('config.json')
```
Comment (gimite) [inline lib/shared/config.rb]: Looks like the path is hard-coded to config.json. Use the parameter of saved_session instead.

File: lib/shared/config.rb [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    end
+
+    def load_config_file
+      json = ::File.open(file_path, 'r').read
```
Comment (gimite) [inline lib/shared/config.rb]: Use ::File.read() instead.

File: README.rdoc (hunk 2 above)
Comment (mtczerwinski) [inline README.rdoc]: :+1:

File: README.rdoc (hunk 3 above)
Comment (mtczerwinski) [inline README.rdoc]: :+1:

**Comment by gimite** [comment]
Thanks! Overall looks good. Made some line comments. Can you review them and fix them?

File: lib/google_drive.rb (hunk 6 above)
Comment (mtczerwinski) [inline lib/google_drive.rb]: Good idea.

File: lib/google_drive.rb (hunk 5 above)
Comment (gimite) [inline lib/google_drive.rb]: Oh can you directly generate this JSON file from Developer Console? That would be nice. Please document how to do it :) But anyway I prefer to hard-code it here because it's only the meaningful value here.

**Comment by mtczerwinski** [comment]
Of course, I will do it as soon as possible. Today, at evening (Polish time) should be prepared fix for it.

**Comment by mtczerwinski** [comment]
Ready

File: README.rdoc [hunk 15]
```diff
@@ -19,22 +19,36 @@ Ver. 1.0.x is not 100% backward compatible with 0.3.x. Some methods have been re
 
 First, follow "Create a client ID and client secret" in {this page}[https://developers.google.com/drive/web/auth/web-server] to get a client ID and client secret for OAuth. Set "Application type" to "Other" in the form to create a client ID if you use GoogleDrive.saved_session method as in the example below.
 
+
+The client ID and client secret you obtained in the step above.
```
Comment (gimite) [inline README.rdoc]: Connection between this sentence and the next sentence is a bit unnatural. Can you simply say e.g.,

> Create config.json file which contains the generated client ID and client secret, which looks like this:

File: README.rdoc [hunk 16]
```diff
@@ -19,22 +19,36 @@ Ver. 1.0.x is not 100% backward compatible with 0.3.x. Some methods have been re
...
+You can keep them in a config file, it should look like this one below:
+
+  {
+    "client_id": "xxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.apps.googleusercontent.com",
+    "client_secret": "xxxxxxxxxxxxxxxxxxxxxxxx",
+    "scope": [
```
Comment (gimite) [inline README.rdoc]: Can you remove "scope" and "refresh_token" from this example? They can be omitted in most cases. You may put a full example in the RDoc of saved_session method, but the example in README should be minimal.

File: lib/google_drive/config.rb [hunk 17]
```diff
@@ -0,0 +1,36 @@
+require 'json'
+
+module GoogleDrive
+  class Config
+    #:nodoc:
```
Comment (gimite) [inline lib/google_drive/config.rb]: This must be in the end of the previous line, I guess:

    class Config #:nodoc:

File: lib/google_drive.rb [hunk 18]
```diff
@@ -86,70 +84,50 @@ def self.login_with_oauth(client_or_access_token, proxy = nil)
...
     #
     # You can specify your own OAuth +client_id+ and +client_secret+. Otherwise the default one is used.
-    def self.saved_session(path = nil, proxy = nil, client_id = nil, client_secret = nil)
+    def self.saved_session(path = './.config.json', proxy = nil, client_id = nil, client_secret = nil)
+      config = Config.new(path)
+
+      if client_id.nil? || client_secret.nil?
+        config.call
+      else
+        config.client_id = client_id
```
Comment (gimite) [inline lib/google_drive.rb]: Even if client_id and client_secret is given, you should still call config.load_config_file because we must use refresh_token in the file if available. Please call config.load_config_file first, then override client_id and client_secret in the parameters.

File: lib/google_drive.rb [hunk 19]
```diff
@@ -86,70 +84,50 @@ def self.login_with_oauth(client_or_access_token, proxy = nil)
     # +path+ defaults to ENV["HOME"] + "/.ruby_google_drive.token".
     #
     # You can specify your own OAuth +client_id+ and +client_secret+. Otherwise the default one is used.
-    def self.saved_session(path = nil, proxy = nil, client_id = nil, client_secret = nil)
+    def self.saved_session(path = './.config.json', proxy = nil, client_id = nil, client_secret = nil)
+      config = Config.new(path)
+
+      if client_id.nil? || client_secret.nil?
+        config.call
```
Comment (gimite) [inline lib/google_drive.rb]: Do you have any reason to alias "load_config_file" as "call"? Can you just call config.load_config_file here directly? "load_config_file" sounds more intuitive than "call".

File: lib/google_drive.rb [hunk 20]
```diff
@@ -86,70 +84,50 @@ def self.login_with_oauth(client_or_access_token, proxy = nil)
...
-          when 2
-            # Old format.
-            token_data = nil
-          else
-            raise(ArgumentError, "Not a token file: %s" % path)
-        end
-      else
-        token_data = nil
-      end
+      token_data = config.refresh_token
```
Comment (gimite) [inline lib/google_drive.rb]: Variable name should be also refresh_token instead of token_data.

File: lib/google_drive/config.rb [hunk 21]
```diff
@@ -0,0 +1,36 @@
+require 'json'
```
Comment (gimite) [inline lib/google_drive/config.rb]: Can you put copyright line here like other files?:

    # Author: Hiroshi Ichikawa <http://gimite.net/>
    # The license of this source is "New BSD Licence"

Please use your name as Author instead.

File: lib/google_drive/config.rb [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+
+module GoogleDrive
+  class Config
+    #:nodoc:
+    attr_accessor :client_id, :client_secret, :scope, :refresh_token, :config_path
+
+    Fields = %w(client_id client_secret scope refresh_token).freeze
+
+    def initialize(config_path)
+      @config_path = ::File.expand_path(config_path)
```
Comment (gimite) [inline lib/google_drive/config.rb]: Do you need to apply expand_path? Please remove it if it's not needed.

File: lib/google_drive/config.rb [hunk 23]
```diff
@@ -0,0 +1,36 @@
+require 'json'
+
+module GoogleDrive
+  class Config
+    #:nodoc:
+    attr_accessor :client_id, :client_secret, :scope, :refresh_token, :config_path
+
+    Fields = %w(client_id client_secret scope refresh_token).freeze
```
Comment (gimite) [inline lib/google_drive/config.rb]: Can you call this FIELDS instead? Capital name is usually for class/module name, I guess.

File: lib/google_drive/config.rb [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+      load_config_file
+    end
+
+    def save
+      ::File.open(config_path, 'w') { |file| file.write(prepare_json) }
+    end
+
+    private
+
+    def prepare_json
```
Comment (gimite) [inline lib/google_drive/config.rb]: "to_json" may be better name.

File: lib/google_drive.rb [hunk 25]
```diff
@@ -86,70 +84,50 @@ def self.login_with_oauth(client_or_access_token, proxy = nil)
...
-        auth.fetch_access_token!()
 
+      auth = client.authorization
+      auth.client_id = config.client_id
+      auth.client_secret = config.client_secret
+      auth.scope = config.scope ||
+        ['https://www.googleapis.com/auth/drive', 'https://spreadsheets.google.com/feeds/']
+      auth.redirect_uri = 'urn:ietf:wg:oauth:2.0:oob'
+
+      unless token_data.empty?
```
Comment (gimite) [inline lib/google_drive.rb]: How about using nil as the default of refresh_token? Then this can be simply:

    if token_data

File: lib/google_drive.rb [hunk 26]
```diff
@@ -86,70 +84,50 @@ def self.login_with_oauth(client_or_access_token, proxy = nil)
...
+        $stderr.print('2. Enter the authorization code shown in the page: ')
+        auth.code = $stdin.gets.chomp
+        auth.fetch_access_token!
+        config.refresh_token = auth.refresh_token
       end
 
-      return GoogleDrive.login_with_oauth(client)
+      config.save
 
+      GoogleDrive.login_with_oauth(client)
```
Comment (gimite) [inline lib/google_drive.rb]: Can you keep "return" keyword here? I usually explicitly return to make it clear that the method returns something.

**Comment by gimite** [comment]
Thanks for quick fixes! Added some more comments.

File: lib/google_drive.rb [hunk 27]
```diff
@@ -86,70 +84,50 @@ def self.login_with_oauth(client_or_access_token, proxy = nil)
     # +path+ defaults to ENV["HOME"] + "/.ruby_google_drive.token".
     #
     # You can specify your own OAuth +client_id+ and +client_secret+. Otherwise the default one is used.
-    def self.saved_session(path = nil, proxy = nil, client_id = nil, client_secret = nil)
+    def self.saved_session(path = './.config.json', proxy = nil, client_id = nil, client_secret = nil)
```
Comment (gimite) [inline lib/google_drive.rb]: Can you keep default to ENV["HOME"] + "/.ruby_google_drive.token" for backward compatibility?

File: lib/google_drive.rb (hunk 20 above)
Comment (mtczerwinski) [inline lib/google_drive.rb]: :+1:

**Comment by mtczerwinski** [comment]
@gimite I have updated and refactored files and I have hope that it is what we want to achieve. If you have any questions, I answer them of course.

**Comment by gimite** [comment]
Thanks! Merged with some clean up.

**Comment by mtczerwinski** [comment]
Superb! Thanks!
